src/runtime_src/core/tools/common/ReportMemory.cpp

```diff
@@ -0,0 +1,447 @@
+}
+
+boost::property_tree::ptree
+populate_memtopology(const xrt_core::device * device, const std::string& desc)
+{
+  boost::property_tree::ptree pt;
+  std::vector<std::string> mm_buf, stream_stat;
+  std::vector<char> buf, temp_buf, gbuf;
+  uint64_t memoryUsage, boCount;
+  pt.put("description", desc);
+  getChannelinfo(device, pt);
+  try {
+    buf = xrt_core::device_query<qr::mem_topology_raw>(device);
+    mm_buf = xrt_core::device_query<qr::memstat_raw>(device);
+    temp_buf = xrt_core::device_query<qr::temp_by_mem_topology>(device);
+  } catch (const std::exception& ex){
+    pt.put("error_msg", ex.what());
+  }
+  
+  if(buf.empty() || mm_buf.empty())
+    return pt;
+
+  const mem_topology *map = (mem_topology *)buf.data();
+  const uint32_t *temp = (uint32_t *)temp_buf.data();
+
+  try {
+    boost::any a = std::string("1");
+    xrt_core::device_update<qr::mig_cache_update>(device,a);
+  } catch (const std::exception& ex){
+    pt.put("error_msg", ex.what());
+  }
+  
+  boost::property_tree::ptree ptMem_array;
+  boost::property_tree::ptree ptStream_array;
+  boost::property_tree::ptree ptGrp_array;
+  for (int i = 0; i < map->m_count; i++) {
+    if (map->m_mem_data[i].m_type == MEM_STREAMING || map->m_mem_data[i].m_type == MEM_STREAMING_CONNECTION) {
+      std::string lname, status = "Inactive", total = "N/A", pending = "N/A";
+      boost::property_tree::ptree ptStream;
+      std::map<std::string, std::string> stat_map;
+      lname = std::string((char *)map->m_mem_data[i].m_tag);
+      ptStream.put("tag", map->m_mem_data[i].m_tag);
+      if (lname.back() == 'w')
+        lname = "route" + std::to_string(map->m_mem_data[i].route_id) + "/stat";
+      else if (lname.back() == 'r')
+        lname = "flow" + std::to_string(map->m_mem_data[i].flow_id) + "/stat";
+      else
+        status = "N/A";
+      try {
+        boost::any a = std::make_pair("entry", lname.c_str());
+        stream_stat = xrt_core::device_query<qr::dma_stream>(device, a);
+        status = "Active";
+        for (unsigned k = 0; k < stream_stat.size(); k++) {
+          std::vector<std::string> strs;
+          boost::split(strs, stream_stat[k],boost::is_any_of(":"));
+          if (strs.size() > 1)
+            stat_map[strs[0]] = strs[1];
+        }
+
+        total = stat_map[std::string("complete_bytes")] + "/" + stat_map[std::string("complete_requests")];
+        pending = stat_map[std::string("pending_bytes")] + "/" + stat_map[std::string("pending_requests")];
+        ptStream.put("usage.status", status);
+        ptStream.put("usage.total", total);
+        ptStream.put("usage.pending", pending);
+      } catch (const std::exception& ){
+        // eat the exception, probably bad path
+      }
+  
+      ptStream_array.push_back(std::make_pair("",ptStream));
+      continue;
+    }
+
+    boost::property_tree::ptree ptMem;
+
+    std::string str = "**UNUSED**";
+    auto search = memtype_map.find((MEM_TYPE)map->m_mem_data[i].m_type );

```

Comment:
No c-style casting, note that `MEM_TYPE(map->m_mem_data[i].m_type) ` would be acceptable.